src/windows.rs

```diff
@@ -1,59 +1,30 @@
 //! Implementation for Windows
 use crate::Error;
-use core::{ffi::c_void, mem::MaybeUninit, num::NonZeroU32, ptr};
+use core::mem::MaybeUninit;
 
-const BCRYPT_USE_SYSTEM_PREFERRED_RNG: u32 = 0x00000002;
+type HMODULE = isize;
+type BOOL = i32;
+const TRUE: BOOL = 1;
 
-#[link(name = "bcrypt")]
-extern "system" {
-    fn BCryptGenRandom(
-        hAlgorithm: *mut c_void,
-        pBuffer: *mut u8,
-        cbBuffer: u32,
-        dwFlags: u32,
-    ) -> u32;
-}
+type ProcessPrng = unsafe extern "system" fn(*mut u8, usize) -> BOOL;
 
-// Forbidden when targetting UWP
-#[cfg(not(target_vendor = "uwp"))]
-#[link(name = "advapi32")]
+#[link(name = "kernel32")]
 extern "system" {
-    #[link_name = "SystemFunction036"]
-    fn RtlGenRandom(RandomBuffer: *mut c_void, RandomBufferLength: u32) -> u8;
+    fn LoadLibraryA(libfilename: *const u8) -> HMODULE;
+    fn GetProcAddress(hmodule: HMODULE, procname: *const u8) -> Option<ProcessPrng>;
 }
 
 pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    // Prevent overflow of u32
-    for chunk in dest.chunks_mut(u32::max_value() as usize) {
-        // BCryptGenRandom was introduced in Windows Vista
-        let ret = unsafe {
-            BCryptGenRandom(
-                ptr::null_mut(),
-                chunk.as_mut_ptr() as *mut u8,
-                chunk.len() as u32,
-                BCRYPT_USE_SYSTEM_PREFERRED_RNG,
-            )
-        };
-        // NTSTATUS codes use the two highest bits for severity status.
-        if ret >> 30 == 0b11 {
-            // Failed. Try RtlGenRandom as a fallback.
-            #[cfg(not(target_vendor = "uwp"))]
-            {
-                let ret =
-                    unsafe { RtlGenRandom(chunk.as_mut_ptr() as *mut c_void, chunk.len() as u32) };
-                if ret != 0 {
-                    continue;
-                }
-            }
-            // We zeroize the highest bit, so the error code will reside
-            // inside the range designated for OS codes.
-            let code = ret ^ (1 << 31);
-            // SAFETY: the second highest bit is always equal to one,
-            // so it's impossible to get zero. Unfortunately the type
-            // system does not have a way to express this yet.
-            let code = unsafe { NonZeroU32::new_unchecked(code) };
-            return Err(Error::from(code));
-        }
+    let dll = unsafe { LoadLibraryA(b"bcryptprimitives.dll\0".as_ptr()) };
```

Comment:
We should use `LoadLibraryeX` with `LOAD_LIBRARY_SEARCH_SYSTEM32` instead (but see the documentation about that flag and compatibility) or `LOAD_WITH_ALTERED_SEARCH_PATH` with an absolute path to the System32 folder.

Google Chrome doesn't security bug reports about this issue to be valid, but not everybody agrees with Google Chrome. Regardless, I can tell you from past experience that this will trigger security vulnerability bug reports to every application that uses getrandom, if we don't restrict the search path to `System32`. These bug reports are difficult for many projects to deal with. It's worth avoiding.

Reply:
If it's in `\KnownDlls` then it doesn't particularly matter because it will indeed prioritize the system directory for those dlls. However `bcryptprimitives` is a bit more awkward because it's a late addition, iirc the earliest versions of Windows 10 did not have it as a known dll.